// code derived from https://github.com/chjj/marked

// https://github.com/chjj/marked/blob/master/LICENSE
/*
Copyright (c) 2011-2014, Christopher Jeffrey (https://github.com/chjj/)

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
*/

diff --git a/lib/marked.js b/lib/marked.js
index 2de4991..2a53e79 100644
--- a/lib/marked.js
+++ b/lib/marked.js
@@ -515,6 +515,7 @@ function InlineLexer(links, options) {
   this.links = links;
   this.rules = inline.normal;
   this.renderer = this.options.renderer || new Renderer;
+  this.renderer.options = this.options;
 
   if (!this.links) {
     throw new
@@ -735,7 +736,9 @@ InlineLexer.prototype.mangle = function(text) {
  * Renderer
  */
 
-function Renderer() {}
+function Renderer(options) {
+  this.options = options || {};
+}
 
 Renderer.prototype.code = function(code, lang, escaped, options) {
   options = options || {};
@@ -845,6 +848,16 @@ Renderer.prototype.del = function(text) {
 };
 
 Renderer.prototype.link = function(href, title, text) {
+  if (this.options.sanitize) {
+    try {
+      var prot = decodeURIComponent(href).replace(/[^\w:]/g, '').toLowerCase();
+    } catch (e) {
+      return '';
+    }
+    if (prot.indexOf('javascript:') === 0) {
+      return '';
+    }
+  }
   var out = '<a href="' + href + '"';
   if (title) {
     out += ' title="' + title + '"';
@@ -872,6 +885,7 @@ function Parser(options) {
   this.options = options || marked.defaults;
   this.options.renderer = this.options.renderer || new Renderer;
   this.renderer = this.options.renderer;
+  this.renderer.options = this.options;
 }
 
 /**
